lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -4932,6 +4930,41 @@ namespace Js
         }
     }
 
+    bool JavascriptOperators::CanShortcutOnUnknownPropertyName(RecyclableObject *instance)
+    {
+        if (!CanShortcutInstanceOnUnknownPropertyName(instance))
+        {
+            return false;
+        }
+        return CanShortcutPrototypeChainOnUnknownPropertyName(instance->GetPrototype());
+    }
+
+    bool JavascriptOperators::CanShortcutInstanceOnUnknownPropertyName(RecyclableObject *instance)
+    {
+        if (!instance->HasDeferredTypeHandler())
+        {
+            return true;
+        }
+        return !(instance->IsExternal() || 
+                 JavascriptProxy::Is(instance) ||
+                 (JavascriptFunction::Is(instance) && JavascriptFunction::FromVar(instance)->IsExternalFunction()));
+    }
+
+    bool JavascriptOperators::CanShortcutPrototypeChainOnUnknownPropertyName(RecyclableObject *prototype)
+    {
+        Assert(prototype);
+
+        TypeId typeId;
```

Comment:
Oh, I see what you’re saying. Sure.

--Paul

From: Ian Halliday [mailto:notifications@github.com]
Sent: Friday, April 8, 2016 2:49 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [Microsoft/ChakraCore] Fix optimized element access that incorrectly reports that the instance has no such property (#723)


In lib/Runtime/Language/JavascriptOperators.cpp<https://github.com/Microsoft/ChakraCore/pull/723#discussion_r59094028>:

> +    bool JavascriptOperators::CanShortcutInstanceOnUnknownPropertyName(RecyclableObject *instance)

> +    {

> +        if (!instance->HasDeferredTypeHandler())

> +        {

> +            return true;

> +        }

> +        return !(instance->IsExternal() ||

> +                 JavascriptProxy::Is(instance) ||

> +                 (JavascriptFunction::Is(instance) && JavascriptFunction::FromVar(instance)->IsExternalFunction()));

> +    }

> +

> +    bool JavascriptOperators::CanShortcutPrototypeChainOnUnknownPropertyName(RecyclableObject *prototype)

> +    {

> +        Assert(prototype);

> +

> +        TypeId typeId;

But it is only assigned in the loop condition and not used later anywhere.

Can't the loop condition change to just prototype->GetTypeId() != TypeIds_Null?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/723/files/43a1f719f0178eedd511ee820762290e0fca4f3f#r59094028>